IdentityCore/src/cache/MSIDMacKeychainTokenCache.m

```diff
@@ -0,0 +1,683 @@
+            }
+        }
+
+        self.keychainGroup = keychainGroup;
+        if (!self.keychainGroup)
+        {
+            return nil;
+        }
+
+        self.defaultAccountQuery = @{
+                                     // All account items are saved as generic passwords.
+                                     (id)kSecClass: (id)kSecClassGenericPassword,
+
+                                     // Add the access group as it's own field for query filtering.
+                                     // Since the attribute is an NSNumber, hash the string.
+                                     (id)kSecAttrCreator: [NSNumber numberWithUnsignedInt:(uint32_t)self.keychainGroup.hash],
+
+                                     // Add a marker for all cache items in the keychain for additional query filtering.
+                                     (id)kSecAttrLabel: s_defaultKeychainLabel
+
+                                     };
+
+        MSID_LOG_INFO(nil, @"Init MSIDMacKeychainTokenCache with keychainGroup: %@", [self keychainGroupLoggingName]);
+        MSID_LOG_INFO_PII(nil, @"Init MSIDMacKeychainTokenCache with keychainGroup: %@", self.keychainGroup);
+    }
+
+    return self;
+}
+
+#pragma mark - Accounts
+
+// Write an account to the macOS keychain cache.
+//
+// Errors:
+// * MSIDErrorDomain/MSIDErrorInternal: invalid key or json deserialization error
+// * MSIDKeychainErrorDomain/OSStatus: Apple status codes from SecItemUpdate()/SecItemAdd()
+//
+- (BOOL)saveAccount:(MSIDAccountCacheItem *)account
+                key:(MSIDCacheKey *)key
+         serializer:(id<MSIDAccountItemSerializer>)serializer
+            context:(id<MSIDRequestContext>)context
+              error:(NSError **)error
+{
+    MSID_TRACE;
+    MSID_LOG_VERBOSE(context, @"Set keychain item, key info (account: %@ service: %@, keychainGroup: %@)", _PII_NULLIFY(key.account), _PII_NULLIFY(key.service), [self keychainGroupLoggingName]);
+    MSID_LOG_VERBOSE_PII(context, @"Set keychain item, key info (account: %@ service: %@, keychainGroup: %@)", key.account, key.service, self.keychainGroup);
+
+    if (!key.service.length)
+    {
+        [self createError:@"Set keychain item with invalid key (service is nil)."
+                domain:MSIDErrorDomain errorCode:MSIDErrorInternal error:error context:context];
+        return NO;
+    }
+
+    if (!key.account.length)
+    {
+        [self createError:@"Set keychain item with invalid key (account is nil)."
+                   domain:MSIDErrorDomain errorCode:MSIDErrorInternal error:error context:context];
+        return NO;
+    }
+
+    NSData *jsonData = [serializer serializeAccountCacheItem:account];
+    if (!jsonData)
+    {
+        [self createError:@"Failed to serialize account to json data."
+                   domain:MSIDErrorDomain errorCode:MSIDErrorInternal error:error context:context];
+        return NO;
+    }
+
+    NSMutableDictionary *query = [self primaryAttributesForKey:key];
```

Comment:
nit: would account and credential keys be different?
In that case, let's call the method primaryAccountAttributesForKey

Reply:
Yes, I believe so, at least partially. We'll look at what's common between them as part of adding credentials and update as appropriate. The name you've suggested seems good.